### PR TITLE
[#12588] Added unit tests for SessionEditFormComponent

### PR DESCRIPTION
**Pull request by kenneySiu**
Part of #12588 

**Outline of Solution**
Added unit tests for the following methods in SessionEditFormComponent
triggerModelChange,
triggerSubmissionOpeningDateModelChange,
courseIdChangeHandler,
submitFormHandler,
cancelHandler,
deleteHandler,
copyHandler,
copyOthersHandler,
closeEditFormHandler

### CODE REVIEW

File: src/web/app/components/session-edit-form/session-edit-form.component.spec.ts [hunk 1]
```diff
@@ -53,4 +57,117 @@ describe('SessionEditFormComponent', () => {
...
+  + 'earliest date and time is earlier than earliest possible time', () => {
+    const date: DateFormat = component.minDateForSubmissionStart;
+    const minTime: TimeFormat = component.minTimeForSubmissionStart;
+    const time: TimeFormat = { hour: minTime.hour - 1, minute: minTime.minute };
+    const configureSubmissionOpeningTimeSpy = jest.spyOn(component, 'configureSubmissionOpeningTime');
+    component.model.submissionStartTime = time;
+    component.triggerSubmissionOpeningDateModelChange(submissionStartDateField, date);
+    component.configureSubmissionOpeningTime(minTime);
+    expect(component.model.submissionStartTime).toStrictEqual(minTime);
+    expect(configureSubmissionOpeningTimeSpy).toHaveBeenCalledWith(minTime);
```
Comment (cedricongjh) [inline src/web/app/components/session-edit-form/session-edit-form.component.spec.ts:70]: we should also expect triggerModelChange to have been called here

File: src/web/app/components/session-edit-form/session-edit-form.component.spec.ts [hunk 2]
```diff
@@ -53,4 +57,117 @@ describe('SessionEditFormComponent', () => {
...
+    expect(cancelEditingSessionSpy).toHaveBeenCalled();
+  });
+
+  it('should emit a deleteExistingSession event after modal confimation', async () => {
+    jest.spyOn((component as any).simpleModalService, 'openConfirmationModal').mockReturnValue(mockModal);
+    const deleteExistingSessionSpy = jest.spyOn(component.deleteExistingSessionEvent, 'emit');
+    component.deleteHandler();
+    await mockModal.result;
+    expect(deleteExistingSessionSpy).toHaveBeenCalled();
+  });
```
Comment (cedricongjh) [inline src/web/app/components/session-edit-form/session-edit-form.component.spec.ts:147]: for these methods, we should be creating a modalSpy and checking if it's called with the appropriate params, you can see admin-notifications-page.component.spec.ts for an example

**Review by cedricongjh — CHANGES_REQUESTED** [review]
some changes,

also minTimeForSubmissionEnd, minDateForSessionVisible, minTimeForSessionVisible, maxDateForSessionVisible, maxTimeForSessionVisible, minDateForResponseVisible, minTimeForResponseVisible are currently not tested as well, do add tests for those

**Comment by kenneySiu** [comment]
Added additional tests 
Ready for review

File: src/web/app/components/session-edit-form/session-edit-form.component.spec.ts [hunk 3]
```diff
@@ -53,4 +67,312 @@ describe('SessionEditFormComponent', () => {
...
+        service.getDateInstance(moment().tz(component.model.timeZone).add(1, 'days'));
+    const maxDateForSessionVisible = component.maxDateForSessionVisible;
+    expect(maxDateForSessionVisible).toEqual(component.model.submissionStartDate);
+  });
+
+  it('should return the customResponseVisibleDate as the maximum date for session visibility '
+      + 'when response visible setting is CUSTOM and submissionStartDate is after customResponseVisibleDate', () => {
+    component.model.responseVisibleSetting = ResponseVisibleSetting.CUSTOM;
+    component.model.submissionStartDate =
+        service.getDateInstance(moment().tz(component.model.timeZone).add(1, 'days'));
```
Comment (cedricongjh) [inline src/web/app/components/session-edit-form/session-edit-form.component.spec.ts:210]: same as above here

File: src/web/app/components/session-edit-form/session-edit-form.component.spec.ts [hunk 4]
```diff
@@ -53,4 +67,312 @@ describe('SessionEditFormComponent', () => {
...
+      + 'when response visible setting is CUSTOM and submissionStartDate is before customResponseVisibleDate', () => {
+    component.model.responseVisibleSetting = ResponseVisibleSetting.CUSTOM;
+    component.model.submissionStartTime =
+        service.getTimeInstance(moment());
+    component.model.customResponseVisibleTime =
+        service.getTimeInstance(moment());
+    component.model.submissionStartDate =
+        service.getDateInstance(moment().tz(component.model.timeZone));
+    component.model.customResponseVisibleDate =
+        service.getDateInstance(moment().tz(component.model.timeZone).add(1, 'days'));
```
Comment (cedricongjh) [inline src/web/app/components/session-edit-form/session-edit-form.component.spec.ts:248]: same here, let's save all the moment() calls into a const instead

File: src/web/app/components/session-edit-form/session-edit-form.component.spec.ts [hunk 5]
```diff
@@ -53,4 +67,312 @@ describe('SessionEditFormComponent', () => {
...
+    component.model.responseVisibleSetting = ResponseVisibleSetting.CUSTOM;
+    component.model.submissionStartTime =
+        service.getTimeInstance(moment());
+    component.model.customResponseVisibleTime =
+        service.getTimeInstance(moment());
+    component.model.submissionStartDate =
+        service.getDateInstance(moment().tz(component.model.timeZone).add(1, 'days'));
+    component.model.customResponseVisibleDate =
+        service.getDateInstance(moment().tz(component.model.timeZone));
+    const maxTimeForSessionVisible = component.maxTimeForSessionVisible;
```
Comment (cedricongjh) [inline src/web/app/components/session-edit-form/session-edit-form.component.spec.ts:264]: same as above here

File: src/web/app/components/session-edit-form/session-edit-form.component.spec.ts [hunk 6]
```diff
@@ -53,4 +67,312 @@ describe('SessionEditFormComponent', () => {
...
+    expect(minDateForResponseVisible).toEqual(component.model.submissionStartDate);
+  });
+
+  it('should return customSessionVisibleDate as the minimum date for response visibility '
+      + 'when session visible setting is CUSTOM', () => {
+    component.model.sessionVisibleSetting = SessionVisibleSetting.CUSTOM;
+    component.model.submissionStartDate =
+        service.getDateInstance(moment().tz(component.model.timeZone));
+    component.model.customSessionVisibleDate =
+        service.getDateInstance(moment().tz(component.model.timeZone));
```
Comment (cedricongjh) [inline src/web/app/components/session-edit-form/session-edit-form.component.spec.ts:288]: we should make customerSessionVisibleDate different from submissionStartDate here

File: src/web/app/components/session-edit-form/session-edit-form.component.spec.ts [hunk 7]
```diff
@@ -53,4 +67,312 @@ describe('SessionEditFormComponent', () => {
...
+    expect(minTimeForResponseVisible).toEqual(component.model.submissionStartTime);
+  });
+
+  it('should return customSessionVisibleTime as the minimum time for response visibility '
+      + 'when session visible setting is CUSTOM', () => {
+    component.model.sessionVisibleSetting = SessionVisibleSetting.CUSTOM;
+    component.model.submissionStartTime =
+        service.getTimeInstance(moment().tz(component.model.timeZone));
+    component.model.customSessionVisibleTime =
+        service.getTimeInstance(moment().tz(component.model.timeZone));
```
Comment (cedricongjh) [inline src/web/app/components/session-edit-form/session-edit-form.component.spec.ts:313]: same here, let's make the customSessionVisibleTime different from submissionStartTime

**Review by cedricongjh — CHANGES_REQUESTED** [review]
thanks for adding the new tests, left a few comments!

**Comment by kenneySiu** [comment]
Ready for review

**Review by cedricongjh — APPROVED** [review]
LGTM! thank you for your work on this!

**Review by weiquu — APPROVED** [review]
LGTM

**Pull request merged**